src/sources/docker.rs

```diff
@@ -263,97 +279,164 @@ impl DockerSource {
         // t2 -- outside: container stoped
         // t3 -- list_containers
         // In that case, logs between [t1,t2] will be pulled to vector only on next start/unpause of that container.
-        let esb = EventStreamBuilder::new(core, out, main_send, shutdown.clone());
+        let esb = EventStreamBuilder {
+            core: Arc::new(core),
+            out,
+            main_send,
+            shutdown: shutdown,
+        };
 
-        // Construct, capture currently running containers, and do main future(self)
         Ok(DockerSource {
             esb,
-            events: Box::new(events) as Box<_>,
+            events: Box::pin(events),
             containers: HashMap::new(),
             main_recv,
             hostname: env::var("HOSTNAME").ok(),
             exclude_self,
-        }
-        .running_containers()
-        // Once this ShutdownSignal resolves it will drop DockerSource and by extension it's ShutdownSignal.
-        .and_then(move |source| source.select(shutdown.map(|_| ())).then(|_| Ok(()))))
+        })
     }
 
     /// Future that captures currently running containers, and starts event streams for them.
-    fn running_containers(mut self) -> impl Future<Item = Self, Error = ()> {
-        let mut options = shiplift::ContainerListOptions::builder();
-
-        // by docker API, using both type of include results in AND between them
-        // TODO: missing feature in shiplift to include ContainerFilter::Name
-
-        // Include-label
-        options.filter(
-            self.esb
-                .core
-                .config
-                .include_labels
-                .clone()
-                .unwrap_or_default()
-                .iter()
-                .map(|s| ContainerFilter::LabelName(s.clone()))
-                .collect(),
-        );
+    async fn handle_running_containers(mut self) -> crate::Result<Self> {
+        let mut filters = HashMap::new();
+
+        // Apply include filters
+        if let Some(include_labels) = &self.esb.core.config.include_labels {
+            filters.insert("label".to_owned(), include_labels.clone());
+        }
+
+        if let Some(include_images) = &self.esb.core.config.include_images {
+            filters.insert("ancestor".to_owned(), include_images.clone());
```

Comment:
I expected too, but `ancestor` works. Probably it's called `ancestor` because container launched on some image and add own layer of file changes, so `ancestor` sounds reasonable.